# Numpy Operations

NumPy is the fundamental package for scientific computing in Python. It is a Python library that provides a multidimensional array object, various derived objects (such as masked arrays and matrices), and an assortment of routines for fast operations on arrays, including mathematical, logical, shape manipulation, sorting, selecting, I/O, discrete Fourier transforms, basic linear algebra, basic statistical operations, random simulation and much more.

- Numpy Array Reshaping
- Numpy Array Sorting
- Numpy Array Splitting
- Numpy Array appending
- Numpy Products

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [1]:
import numpy as np

## Function 1 - Numpy Array Reshaping(np.reshape)

Reshaping arrays:
 - Reshaping means changing the shape of an array.
 - The shape of an array is the number of elements in each dimension.
 - By reshaping we can add or remove dimensions or change number of elements in each dimension.


In [2]:
# Example 1 working

ar1 = np.arange(8)
print(ar1,'\n')
ar2 = np.arange(8).reshape(2, 4)
print(ar2)

[0 1 2 3 4 5 6 7] 

[[0 1 2 3]
 [4 5 6 7]]


In this example we have created 1-dimensional array and stored in a variable called `ar1`, and by using `.reshape()` function we are changing it to 2-dimensional array and created 2*4 matrix.

In [3]:
# Example 2 working

ar3 = np.arange(8)
print(ar3,'\n')
ar4 = np.arange(8).reshape(2, 2, 2)
print(ar4)

[0 1 2 3 4 5 6 7] 

[[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


In this example we have created 1-dimensional array and stored in a variable called ar1, and by using `.reshape()` function we are changing it to 2-dimensional array .

In [21]:
# Example 3 breaking

ar5 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])
print(ar5,'\n')
ar6 = ar5.reshape(3, 3)
print(ar6)

[ 1  2  3  4  5  6  7  8  9 10 11 12] 



ValueError: cannot reshape array of size 12 into shape (3,3)

In this example we are creating we creating one dimensional array and we trying to reshape the with 3 * 3 matrix ,As you can see it cannot change its shape if you want to change the shape you can try with 4*3 matrix.

## Function 2 - Numpy Array Sorting(np.sort())

Sorting Arrays:
 - Sorting means putting elements in an ordered sequence.
 - Ordered sequence is any sequence that has an order corresponding to elements, like numeric or alphabetical, ascending or descending.
 - The `NumPy ndarray` object has a function called sort(), that will sort a specified array.

In [22]:
# Example 1 - working

arr = np.array([3, 2, 0, 1])

print(np.sort(arr))

[0 1 2 3]


In this example we have created the array of integers and stored the values randomly, Now we are applying the `.sort()` function so that the array of elements will be sorted in ascending order.

In [23]:
# Example 2 - working

arr = np.array(['banana', 'cherry', 'apple'])

print(np.sort(arr))

['apple' 'banana' 'cherry']


In this example we are creating the array of string values , if we apply the function `np.sort()` the array order will be sorted alphabetically. 

In [24]:
# Example 3 - Working

arr = np.array([[3, 2, 4], [5, 0, 1]])

print(np.sort(arr))

[[2 3 4]
 [0 1 5]]


In this example we are sorting the 2-dimensional array by using `np.sort()` function.

## Function 3 - Numpy Array Splitting(array_split)

Splitting Numpy Arrays:
 - Splitting is reverse operation of Joining.
 - Joining merges multiple arrays into one and Splitting breaks one array into multiple.
 - We use array_split() for splitting arrays, we pass it the array we want to split and the number of splits.

In [25]:
# Example 1 - working

a = np.array([1,2,3,4,5,6])
b = np.array_split(a, 2)
print(b)

[array([1, 2, 3]), array([4, 5, 6])]


In the given example we have created 1-dimensional array and applying the `array_split()` function so that the array could be split and make it 2 dimensional array

In [26]:
# Example 2 - working

c = np.array([[1,2,3], [4,5,6], [7,8,9], [10,11,12], [13,14,15], [16,17,18]])
d = np.array_split(c, 2)
print(d)

[array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]]), array([[10, 11, 12],
       [13, 14, 15],
       [16, 17, 18]])]


In the given example of array we just applied `np.array_split` function and you can see the arrays are has been changed to 2-d array.

In [27]:
# Example 3 - Working

c = np.array([[1,2,3], [4,5,6], [7,8,9], [10,11,12], [13,14,15], [16,17,18]])
d = np.array_split(c, 2, axis=1)
print(d)

[array([[ 1,  2],
       [ 4,  5],
       [ 7,  8],
       [10, 11],
       [13, 14],
       [16, 17]]), array([[ 3],
       [ 6],
       [ 9],
       [12],
       [15],
       [18]])]


In the given example of array we just applied `np.array_split` function and you can see the arrays are has been changed to 2-d array, and we assigned axis=1 so that new columns in a array has been created.

## Function 4 - Numpy Array appending(np.append


The numpy.append() appends values along the mentioned axis at the end of the array


In [28]:
# Example 1 - working

arr1 = np.arange(8).reshape(2, 4)
print(arr1,'\n')
arr2 = np.arange(8, 16).reshape(2, 4)
print(arr2,'\n')
arr3 = np.append(arr1, arr2, axis=0)
print(arr3)

[[0 1 2 3]
 [4 5 6 7]] 

[[ 8  9 10 11]
 [12 13 14 15]] 

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


In this given example we have created two 2-d arrays and applying the `np.append()` function so that the values in the arrays will be combined and appended in the single array.

In [29]:
# Example 2 - working

arr4 = np.arange(8).reshape(4, 2)
print(arr4,'\n')
arr5 = np.arange(8, 16).reshape(4, 2)
print(arr5,'\n')
arr6 = np.append(arr4, arr5)
print(arr6)

[[0 1]
 [2 3]
 [4 5]
 [6 7]] 

[[ 8  9]
 [10 11]
 [12 13]
 [14 15]] 

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]


In this given example we have created two of 2-d arrays and applying the `np.append()` function so that the values in the arrays will be combined and appended in the single array in 1-dimensional.

In [30]:
# Example 3 - breaking

arr4 = np.arange(8).reshape(4, 2)
print(arr4,'\n')
arr5 = np.arange(8, 16).reshape(4, 2)
print(arr5,'\n')
arr6 = np.append(arr4, arr5, arr1, arr2)
print(arr6)

[[0 1]
 [2 3]
 [4 5]
 [6 7]] 

[[ 8  9]
 [10 11]
 [12 13]
 [14 15]] 



TypeError: _append_dispatcher() takes from 2 to 3 positional arguments but 4 were given

In this given example we have created two arrays and applying the `np.append()` function so that the values in the arrays will be combined and appended the single array and as you can see that it is throwing an error because append function allows you to pass two arguments but given 4, if you you give two arguments the append function will work.

## Function 5 - Numpy Products(prod())

To find the product of the elements in an array, use the prod() function.

In [31]:
# Example 1 - working

arr = np.array([1, 2, 3, 4])

x = np.prod(arr)

print(x)

24


Returns: 24 because 1*2*3*4 = 24

In [32]:
# Example 2 - working

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

x = np.prod([arr1, arr2])

print(x)

40320


Returns: 40320 because 1*2*3*4*5*6*7*8 = 40320

In [33]:
# Example 3 - Working

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

newarr = np.prod([arr1, arr2], axis=1)

print(newarr)

[  24 1680]


If you specify axis=1, NumPy will return the product of each array.

Returns: [24 1680]

## Conclusion

In this Notebook we have covered few Numpy functions, we have covered about some numpy creation and manipulation for each function there are some examples given.Hence we covered 
 - Numpy Array Reshaping
 - Numpy Array Sorting
 - Numpy Array Splitting
 - Numpy Array appending
 - Numpy Products.

## Reference Links

* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* Numpy W3-Schools tutorial : https://docs.scipy.org/doc/numpy-dev/